### PR TITLE
docs(argo-workflows): Use helm docs for workflows

### PR DESCRIPTION
**Pull request by tuananh**
Related issue: #696 

add a step to check if docs is staled and need update. it looks like this

image

if docs is up to date, it will look like this

image

Signed-off-by: Tuan Anh Tran <me@tuananh.org>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] I have bumped the chart version according to [versioning](https://github.com/argoproj/argo-helm/blob/master/CONTRIBUTING.md#versioning)
* [x] I have updated the chart changelog with all the changes that come with this pull request according to [changelog](https://github.com/argoproj/argo-helm/blob/master/CONTRIBUTING.md#changelog).
* [x] Any new values are backwards compatible and/or have sensible default.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo).
* [ ] My build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)).

Changes are automatically published when merged to `master`. They are not published on branches.

### CODE REVIEW

File: charts/argo-workflows/values.yaml [hunk 1]
```diff
@@ -248,50 +248,48 @@ server:
...
-  ## Additional volumes to the server main container.
+  # -- Additional volumes to the server main container.
   volumeMounts: []
   volumes: []
 
-  ## Ingress configuration.
-  ## ref: https://kubernetes.io/docs/user-guide/ingress/
-  ##
+  # -- Ingress configuration.
+  # ref: https://kubernetes.io/docs/user-guide/ingress/
```
Comment (vladlosev) [inline charts/argo-workflows/values.yaml:283]: Putting the comment here appears to generate a huge default value in the output table, which is hard to parse. Please remove it here and add comments to the individual fields instead.

**Review by vladlosev — COMMENTED** [review]
This is a very nice change, thanks! But as it's submitted I see a big maintenance issue: not all people making changes to the chart will be commenting the values in the proper way or running the helm-docs before each commit. Eventually the `README.md` will go out seriously of sync with `values.yaml`. So in order for this to be a viable change going forward, two things need to be done:
- Give users an easy way to re-generate `README.md` when updating the values `values.yaml`.
- Require the users to do that when submitting the pull request.

The first requirement can be fulfilled by providing a script to bootstrap and invoke helm-docs for a given chart. The second on can met by amending the CI configuration by adding a step that runs helm-docs on charts that have `README.md.gotmpl` provided, comparing results with the existing `README.md`, and failing if there is a mismatch. This will force people to make sure they always run helm-docs before submitting the PR.

File: charts/argo-workflows/values.yaml (hunk 1 above)
Comment (tuananh) [reply charts/argo-workflows/values.yaml:283]: i will add a step in CI to check whether the readme is stale and need to re-generate

Comment (mkilchhofer) [reply charts/argo-workflows/values.yaml:283]: Hi @tuananh. Thanks for this great contribution. I like this and I also converted several charts to use helm-docs here.
Can you prefix the unnecessary comment line with two `##` so it doesn't get included in the README? This way we can get rid of the long and unnecessary information in the README.

Comment (tuananh) [reply charts/argo-workflows/values.yaml:283]: @mkilchhofer updated. i also add a step in CI to check if the docs is up to date or not.

File: .github/workflows/lint-and-test.yml [hunk 2]
```diff
@@ -35,6 +35,12 @@ jobs:
           fi
       - name: Run chart-testing (lint)
         run: ct lint --debug --config ./.github/configs/ct-lint.yaml --lint-conf ./.github/configs/lintconf.yaml
+      
+      - name: Run docs-testing (helm-docs)
+        uses: buttahtoast/helm-release-action@v2.0.1
+        with:
+          charts: "${{ steps.list-changed.outputs.changed_charts }}"
```
Comment (vladlosev) [inline .github/workflows/lint-and-test.yml:42]: As far as I can tell, this will pick up and try to compare the results of`helm-docs` on all the charts, not just the ones that are helm-docs enabled (have `README.md.gotmpl`). That will break the build on a change to any chart that is not yet enabled for helm-docs. One way around it is to generate a separate list of changed charts with helm-docs support and run the check only on them. For example:
```bash
          changed=$(ct --config ./.github/configs/ct-lint.yaml list-changed)
          charts=$(echo "$changed" | tr '\n' ' ' | xargs)
          if [[ -n "$changed" ]]; then
            echo "::set-output name=changed::true"
            echo "::set-output name=changed_charts::$charts"
          fi
          changed_helm_docs=$(for chart in $changed ; do if test -f "$chart/README.md.gotmpl" ; then echo "$chart" ; done)
          charts_helm_docs=$(echo "$changed_helm_docs" | tr '\n' ' ' | xargs)
          if test -n "$changed_helm_docs" ; then
            echo "::set-output name=changed_helm_docs::true"
            echo "::set-output name=changed_helm_docs_charts::$charts_helm_docs"
          fi
```

**Review by vladlosev — COMMENTED** [review]
One other issue is you need to provide or document an easy way to setup and run `helm-docs`. Without that many people will be avoid either updating values.yaml or making changes to the chart altogether.

File: .github/workflows/lint-and-test.yml (hunk 2 above)
Comment (mkilchhofer) [reply .github/workflows/lint-and-test.yml:42]: We could place the same note like it exists in the image-updaters README

**Comment by mkilchhofer** [comment]
Ah I will add a commit with the remaining (undocumented/missing) values asap

File: .github/workflows/lint-and-test.yml (hunk 2 above)
Comment (mkilchhofer) [reply .github/workflows/lint-and-test.yml:42]: Add the remaining charts to a file .helmdocsignore?

Comment (mkilchhofer) [reply .github/workflows/lint-and-test.yml:42]: I added the .helmdocsignore. Are you okay with that, @vladlosev ?

Comment (vladlosev) [reply .github/workflows/lint-and-test.yml:42]: Yes, looks like it will work. The only downside is the check will not be automatically enabled for a chart if one adds the template there - one will also have to remember (or to know) to delete the directory from .helmdocsignore.

Comment (mkilchhofer) [reply .github/workflows/lint-and-test.yml:42]: Okay but in my eyes this sounds reasonable as such a contributor will be familiar with helm-docs.

Comment (vladlosev) [reply .github/workflows/lint-and-test.yml:42]: I'll trust your opinion here, then.

**Review by mkilchhofer — APPROVED** [review]
